packages/truffle-core/lib/debug/interpreter.js

```diff
@@ -0,0 +1,537 @@
+const debugModule = require("debug");
+const debug = debugModule("lib:debug:interpreter");
+
+const path = require("path");
+const util = require("util");
+const ora = require("ora");
+
+const DebugUtils = require("truffle-debug-utils");
+const selectors = require("truffle-debugger").selectors;
+const { solidity, trace, controller } = selectors;
+
+const analytics = require("../services/analytics");
+const ReplManager = require("../repl");
+
+const { DebugPrinter } = require("./printer");
+
+function watchExpressionAnalytics(raw) {
+  if (raw.includes("!<")) {
+    //don't send analytics for watch expressions involving selectors
+    return;
+  }
+  let expression = raw.trim();
+  //legal Solidity identifiers (= legal JS identifiers)
+  let identifierRegex = /^[a-zA-Z_$][a-zA-Z_$0-9]*$/;
+  let isVariable = expression.match(identifierRegex) !== null;
+  analytics.send({
+    command: "debug: watch expression",
+    args: { isVariable }
+  });
+}
+
+class DebugInterpreter {
+  constructor(config, session, txHash) {
+    this.config = config;
+    this.session = session;
+    this.printer = new DebugPrinter(config, session);
```

Comment:
Just a general comment about the whole `DebugPrinter` idea -- it's kind of confusing that, even though you have this debug printer, much of the printing doesn't go through it, and is just raw `config.logger.log`.  I mean, I dunno that it makes sense to write a whole new method for every message we want to print, but maybe at least the `DebugPrinter` could get a simple `print` method that all these uses of `config.logger.log` could be replaced with?  So that at least all printing would go through the printer?

Reply:
I recognized the lack of cohesion as I was putting `DebugPrinter` together, but I decided that it could be a separate effort to clean up the input/output. My thinking is that this PR is mostly a gentle structural nudge, not a full clean-up. Moving around actual `config.logger` statements seems like it's bound to result in slip-ups that are hard to account for.

We'll want to do that eventually, but it's just a question of whether it belongs in this PR. My feeling is that it's not urgent enough to block this PR, but let me know if you disagree. Your point may be sound: it might not be worth making this class without it being more cohesive and decoupled. What do you think?

Reply:
I mean it shouldn't be hard to do a search/replace to turn `config.logger.log` into `this.printer.print` should it?  Obviously that's kind of a stopgap, but I think it's a lot better than raw calls to `config.logger.log`.

Reply:
Oh I see. Yeah I'll do that!

Reply:
OK I got rid of the `this.config.logger.log` statements in the interpreter, and even went a bit further there to get rid of `this.config`, since it was barely used.